proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BungeeCordMessageResponder.java

```diff
@@ -0,0 +1,325 @@
+
+      sendResponse(buf);
+    });
+  }
+
+  private void processKick(ByteBufDataInput in) {
+    proxy.getPlayer(in.readUTF()).ifPresent(player -> {
+      String kickReason = in.readUTF();
+      player.disconnect(LegacyComponentSerializer.INSTANCE.deserialize(kickReason));
+    });
+  }
+
+  private ByteBuf prepareForwardMessage(ByteBufDataInput in) {
+    String channel = in.readUTF();
+    short messageLength = in.readShort();
+
+    ByteBuf buf = Unpooled.buffer();
+    ByteBufDataOutput forwarded = new ByteBufDataOutput(buf);
+    forwarded.writeUTF(channel);
+    forwarded.writeShort(messageLength);
+    buf.writeBytes(in.unwrap().readSlice(messageLength));
+    return buf;
+  }
+
+  private void processForwardToPlayer(ByteBufDataInput in) {
+    proxy.getPlayer(in.readUTF())
+        .flatMap(Player::getCurrentServer)
+        .ifPresent(server -> sendResponse(player, prepareForwardMessage(in)));
+  }
+
+  private void processForwardToServer(ByteBufDataInput in) {
+    ByteBuf toForward = prepareForwardMessage(in);
+    String target = in.readUTF();
+    if (target.equals("ALL")) {
+      ByteBuf unreleasableForward = Unpooled.unreleasableBuffer(toForward);
+      try {
+        for (RegisteredServer rs : proxy.getAllServers()) {
+          ((VelocityRegisteredServer) rs).sendPluginMessage(LEGACY_CHANNEL, unreleasableForward);
+        }
+      } finally {
+        toForward.release();
+      }
+    } else {
+      proxy.getServer(target).ifPresent(rs -> ((VelocityRegisteredServer) rs)
+          .sendPluginMessage(LEGACY_CHANNEL, toForward));
+    }
+  }
+
+  // Note: this method will always release the buffer!
+  private void sendResponse(ByteBuf buf) {
+    sendResponse(this.player, buf);
+  }
+
+  // Note: this method will always release the buffer!
+  private static void sendResponse(ConnectedPlayer player, ByteBuf buf) {
+    MinecraftConnection serverConnection = player.ensureAndGetCurrentServer().ensureConnected();
+    String chan = serverConnection.getProtocolVersion().compareTo(ProtocolVersion.MINECRAFT_1_13)
+        >= 0 ? MODERN_CHANNEL.getId() : LEGACY_CHANNEL.getId();
+
+    PluginMessage msg = null;
+    boolean released = false;
+
+    try {
+      VelocityServerConnection vsc = player.getConnectedServer();
+      if (vsc == null) {
+        return;
+      }
+
+      MinecraftConnection serverConn = vsc.ensureConnected();
+      msg = new PluginMessage(chan, buf);
+      serverConn.write(msg);
+      released = true;
+    } finally {
+      if (!released && msg != null) {
+        msg.release();
+      }
+    }
+  }
+
+  public boolean process(PluginMessage message) {
+    if (!MODERN_CHANNEL.getId().equals(message.getChannel()) && !LEGACY_CHANNEL.getId()
+        .equals(message.getChannel())) {
+      return false;
+    }
+
+    ByteBufDataInput in = new ByteBufDataInput(message.content());
+    String subChannel = in.readUTF();
+    if (subChannel.equals("ForwardToPlayer")) {
```

Comment:
A `switch` would be better here

Reply:
The subchannel handling is getting split up anyway, so this is effectively redundant.